apps/internal/base/base.go

```diff
@@ -296,38 +300,21 @@ func (b Client) AcquireTokenSilent(ctx context.Context, silent AcquireTokenSilen
 	authParams.Claims = silent.Claims
 	authParams.UserAssertion = silent.UserAssertion
 
-	var storageTokenResponse storage.TokenResponse
-	if authParams.AuthorizationType == authority.ATOnBehalfOf {
-		if s, ok := b.pmanager.(cache.Serializer); ok {
-			suggestedCacheKey := authParams.CacheKey(silent.IsAppCache)
-			err = b.cacheAccessor.Replace(ctx, s, cache.ReplaceHints{PartitionKey: suggestedCacheKey})
-			if err != nil {
-				return ar, err
-			}
-			defer func() {
-				err = b.export(ctx, s, suggestedCacheKey, err)
-			}()
-		}
-		storageTokenResponse, err = b.pmanager.Read(ctx, authParams)
-		if err != nil {
-			return ar, err
-		}
-	} else {
-		if s, ok := b.manager.(cache.Serializer); ok {
-			suggestedCacheKey := authParams.CacheKey(silent.IsAppCache)
-			err = b.cacheAccessor.Replace(ctx, s, cache.ReplaceHints{PartitionKey: suggestedCacheKey})
-			if err != nil {
-				return ar, err
-			}
-			defer func() {
-				err = b.export(ctx, s, suggestedCacheKey, err)
-			}()
-		}
+	m := b.pmanager
+	if authParams.AuthorizationType != authority.ATOnBehalfOf {
 		authParams.AuthorizationType = authority.ATRefreshToken
-		storageTokenResponse, err = b.manager.Read(ctx, authParams, silent.Account)
-		if err != nil {
-			return ar, err
-		}
+		m = b.manager
+	}
+	key := authParams.CacheKey(silent.IsAppCache)
+	b.mu.RLock()
```

Comment:
Shouldn't this be the write lock?

Reply:
It doesn't need to be, or at least I haven't come up with a scenario requiring an exclusive lock here :thinking:. Here's my reasoning: `Replace` unmarshals persisted data into the in-memory cache, so calling it repeatedly can cause the in-memory cache to lose only data that's been removed from the persistent cache. That could happen and may be a bug, but either way it's outside the responsibility of this type, which just needs to prevent its methods from concurrently calling `Replace` and `Export`.

I'd like to read lock here, provided that's safe, because this is a hot path. I haven't benchmarked this but I expect a significant perf advantage for clients without a persisted cache (i.e. almost all of them) because `Replace` is a no-op for them.

Reply:
The persisted cache can lock to prevent simultaneous Export and Replace but only for the duration of each call. That approach can't prevent the destructive interleaved calls I described above. Replace->authenticate->Export is a transaction only at a higher level.

There's another way to eliminate the locking overhead for clients without a persistent cache. Currently, the client has a default "persistent" cache with no-op Export/Replace and always locks around them. If the default were `nil` instead, the client could decide to lock only when it's not `nil` i.e. the user specified a persistent cache.